### PR TITLE
🪙 [Experimental] Support GSPO-token

### PR DESCRIPTION
**Pull request by hjh0119**
Support for GSPO-token as described in [GSPO paper](https://arxiv.org/abs/2507.18071), Section 4.3.

related issue: https://github.com/huggingface/trl/issues/3811

**GSPO** 
$w_{i}^{\mathrm{GSPO}} = \left[ \frac{\pi_{\theta}(y_i \mid x)}{\pi_{\theta_{\mathrm{old}}}(y_i \mid x)} \right]^{\frac{1}{|y_i|}} = \exp(\frac{1}{|y_i|} \sum_{t=1}^{|y_i|} \log \frac{\pi_{\theta}(y_{i, t} \mid x, y_{i, <t})}{\pi_{\theta_{\mathrm{old}}}(y_{i, t} \mid x, y_{i, <t})})$

**GSPO-token**
$w_{i, t}^{\mathrm{GSPO_token}} = \mathrm{sg}\left[w_i^{\mathrm{GSPO}}\right] \cdot \frac{\pi_{\theta}(y_{i, t} \mid x, y_{i, < t})}{\mathrm{sg}\left[\pi_{\theta}(y_{i, t} \mid x, y_{i, < t})\right]}$

> where $\mathrm{sg}[\cdot]$ denotes the stop-gradient (detach) operation.

> 💡 **NOTE:** GSPO-token enables support for fine-grained (token-level) advantages. 
> However, given the current formulation for advantage computation, all tokens within a sentence share the same value. In this case, GSPO and GSPO-token are theoretically equivalent, as shown in equations (11) and (18) of the paper.

### CODE REVIEW

**Comment by LeonEricsson** [comment]
Thanks for this. Since GSPO-token is a generalized version of vanilla GSPO, I suggest we fully transition to GSPO-token instead of supporting both versions. Consequently, we would rename/remove `importance_sampling_level`, as both methods operate at the token level.

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1837,6 +1837,11 @@ def _compute_loss(self, model, inputs):
         elif self.importance_sampling_level == "sequence":
             log_importance_weights = (log_ratio * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)
             log_importance_weights = log_importance_weights.unsqueeze(-1)
+        elif self.importance_sampling_level == 'sequence_token':
+            # GSPO-token: sg[si(θ)] * πθ(yi,t)/sg[πθ(yi,t)]
+            seq_level_log_weight = (log_ratio * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)
+            seq_level_log_weight = seq_level_log_weight.unsqueeze(-1).detach()  # Stop gradient
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
            seq_level_log_weight = seq_level_log_weight.detach().unsqueeze(-1) # Stop gradient
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1837,6 +1837,11 @@ def _compute_loss(self, model, inputs):
...
+            log_importance_weights = seq_level_log_weight + per_token_logps - per_token_logps.detach()
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
            log_importance_weights = per_token_logps - per_token_logps.detach() + seq_level_log_weight
```
reorder to avoid precision errors

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: `(log_ratio * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)`

This op is common across GSPO and GSPO-token, would be good to have a single variable pointing to this value under an if condition like

`if self.importance_sampling_level != 'token'`

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (hjh0119) [reply trl/trainer/grpo_trainer.py]: It should still be float32 after reordering

So maybe we need to manually convert the dtype like

```python
log_importance_weights = (seq_level_log_weight + per_token_logps - \
per_token_logps.detach()).to(per_token_logps.dtype)
```

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (hjh0119) [reply trl/trainer/grpo_trainer.py]: make sense, so shall we move the invalid value check for importance_sampling_level into the model parameter initialization?

**Comment by hjh0119** [comment]
> Thanks for this. Since GSPO-token is a generalized version of vanilla GSPO, I suggest we fully transition to GSPO-token instead of supporting both versions. Consequently, we would rename/remove `importance_sampling_level`, as both methods operate at the token level.

Agreed to keep GSPO-token. Should we retain this parameter for compatibility with previous usage, or introduce an additional parameter instead? Which is better?

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: wasn't referring to the dtype, just better to perform the subtraction `per_token_logps - per_token_logps.detach()` first, to avoid catastrophic cancellation

Comment (hjh0119) [reply trl/trainer/grpo_trainer.py]: that make sense, fixed

**Comment by hjh0119** [comment]
@qgallouedec @lewtun @edbeeching @kashif If there are any concerns or suggestions, please feel free to let me know. Thank you very much in advance

**Comment by LeonEricsson** [comment]
> Agreed to keep GSPO-token. Should we retain this parameter for compatibility with previous usage, or introduce an additional parameter instead? Which is better?

imo it should be removed, however, since it's already been published as part of TRL v0.20, we may need to keep it for backward comp. I can't speak to it myself, so I'll leave it to someone else to decide.

**Comment by qgallouedec** [comment]
Thanks for the contribution, and apologies for the delay in reviewing it.  

After reading the paper, I don’t think this PR fully achieves GSPO-token. This variation is most relevant when you have a fine-grained advantage— i.e. when $\hat{A_{i,t}}$ varies with $t$—which isn't the case here, since $\hat{A_{i,t}}=\hat{A_{i}}$.

**Comment by hjh0119** [comment]
> After reading the paper, I don’t think this PR fully achieves GSPO-token. This variation is most relevant when you have a fine-grained advantage— i.e. when A i , t ^ varies with t —which isn't the case here, since A i , t ^ = A i ^ .

Sure. As I mentioned, when there is no fine-grained advantage, the gspo-token gradient is equivalent to the original implementation. 

However, do we need to implement this algorithm in advance to accommodate possible future fine-grained advantage, or to make it easier for downstream users to implement their own customized fine-grained advantage? Anyway, thank you for your review.

**Comment by hjh0119** [comment]
Any feedback before I close this PR, or should we go ahead and merge it?

**Comment by qgallouedec** [comment]
Please leave it open, we are working hard to provide fast review for all the PRs 🙏

**Review by qgallouedec — APPROVED** [review]
